internal/cosesign1/check.go

```diff
@@ -0,0 +1,282 @@
+	for i, cert := range chain {
+		if i == 0 {
+			leafCert = cert
+			if verbose {
+				log.Println(x509ToPEM(cert))
+			}
+		} else if i == chainLen-1 {
+			rootCert = cert
+			rootCerts.AddCert(rootCert)
+			if verbose {
+				log.Println(x509ToPEM(cert))
+			}
+		} else {
+			intermediateCerts.AddCert(cert)
+			if verbose {
+				log.Println(x509ToPEM(cert))
+			}
+		}
+	}
+
+	opts := x509.VerifyOptions{
+		Intermediates: intermediateCerts,
+		Roots:         rootCerts,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny}, // To be removed when I have a decent set of examples.
+		// consider CurrentTime time.Time set to be the time the cose message was signed so we are checking the certs were valid then rather than now. Maybe not TBD
+	}
+
+	/*
+		Until we have some production certs allow the certificate check to fail
+	*/
+
+	_, err = leafCert.Verify(opts)
+
+	if err != nil {
+		return nil, fmt.Errorf("certificate chain verification failed")
+	}
+
+	var leafCertBase64 = x509ToBase64(leafCert) // blob of the leaf x509 cert reformatted into pem (base64) style as per the fragment policy rules expect
+	var leafPubKey = leafCert.PublicKey
+	var leafPubKeyBase64 = keyToBase64(leafPubKey)
+
+	var chainPEM string
+	for i, c := range chain {
+		if i > 0 {
+			chainPEM += "\n"
+		}
+		chainPEM += x509ToPEM(c)
+	}
+
+	var results = UnpackedCoseSign1{
+		Pubcert:     leafCertBase64,
+		Feed:        feed,
+		Issuer:      issuer,
+		Pubkey:      leafPubKeyBase64,
+		ChainPem:    chainPEM,
+		ContentType: msg.Headers.Protected[cose.HeaderLabelContentType].(string),
+		Payload:     msg.Payload,
+		CertChain:   chain,
+	}
+
+	if err != nil {
+		if verbose {
+			log.Print("leafCert.Verify failed: " + err.Error())
+		}
+		return nil, err
+	}
+
+	// Use the supplied public key or the one we extracted from the leaf cert.
+	var keyToCheck any
+	if len(optionalPubKeyPEM) == 0 {
+		keyToCheck = leafPubKey
+	} else {
+		var keyDer *pem.Block
+		keyDer, _ = pem.Decode(optionalPubKeyPEM) // _ is the remaining. We only care about the first key.
+		var keyBytes = keyDer.Bytes
+
+		keyToCheck, err = x509.ParsePKCS1PublicKey(keyBytes)
+		if err == nil {
+			if verbose {
+				log.Printf("parsed as PKCS1 public key %q\n", keyToCheck)
+			}
+		} else {
+			keyToCheck, err = x509.ParsePKIXPublicKey(keyBytes)
+			if err == nil {
+				if verbose {
+					log.Printf("parsed as PKIX key %q\n", keyToCheck)
+				}
+			} else {
+				if verbose {
+					log.Print("Failed to parse provided public key - Error = " + err.Error())
+				}
+				return nil, err
+			}
+		}
+	}
+
+	verifier, err := cose.NewVerifier(algo, keyToCheck)
+	if err != nil {
+		if verbose {
+			log.Printf("cose.NewVerifier failed (algo %d): %s", algo, err.Error())
+		}
+		return nil, err
+	}
+
+	err = msg.Verify(nil, verifier)
+	if err != nil {
+		if verbose {
+			log.Printf("msg.Verify failed: algo = %d err = %s", algo, err.Error())
+		}
+		return nil, err
+	}
+
+	return &results, nil
```

Comment:
Does it make sense to build the result object here instead of above the verification?  In the many return cases, the results would be created and never used or logged.

Reply:
OK.

An alternative (which is how it used to work) would be to return the object in the failure cases too so that calling code can have a rummage to see what was inside the offending blob.

Reply:
I like that pattern as well.  There are many examples in the built-in libraries that return both data and an error.  I noticed that returning objects and errors aren't really a pattern in this repo, but it would make sense to start doing it.